gapis/api/vulkan/overdraw.go

```diff
@@ -0,0 +1,1636 @@
+// Copyright (C) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vulkan
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/gapid/core/memory/arena"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/transform"
+	"github.com/google/gapid/gapis/capture"
+	"github.com/google/gapid/gapis/memory"
+	"github.com/google/gapid/gapis/messages"
+	"github.com/google/gapid/gapis/replay"
+	"github.com/google/gapid/gapis/service"
+	"github.com/google/gapid/gapis/service/path"
+)
+
+var (
+	_ = transform.Transformer(&stencilOverdraw{})
+)
+
+type stencilOverdraw struct {
+	rewrite    map[api.CmdID]replay.Result
+	lastSubIdx map[api.CmdID]api.SubCmdIdx
+}
+
+func newStencilOverdraw() *stencilOverdraw {
+	return &stencilOverdraw{
+		rewrite:    map[api.CmdID]replay.Result{},
+		lastSubIdx: map[api.CmdID]api.SubCmdIdx{},
+	}
+}
+
+func (s *stencilOverdraw) add(ctx context.Context, extraCommands uint64, after []uint64, capt *path.Capture, res replay.Result) {
+	c, err := capture.ResolveFromPath(ctx, capt)
+	if err != nil {
+		res(nil, err)
+	}
+	lastSubmit := int64(after[0])
```

Comment:
Seems like this could be refactored into something more readable:

```go
for lastSubmit := int64(after[0]); lastSubmit >= 0; lastSubmit-- {
	switch (c.Commands[lastSubmit]).(type) {
	case *VkQueueSubmit:
		id := api.CmdID(uint64(lastSubmit) + extraCommands)
		s.rewrite[id] = res
		s.lastSubIdx[id] = api.SubCmdIdx(after[1:])
		return
	}
}
res(nil, &service.ErrDataUnavailable{
	Reason: messages.ErrMessage("No last queue submission"),
})
```